src/noise/rate_limiter.rs

```diff
@@ -0,0 +1,216 @@
+    mac1_key: [u8; 32],
+    mac2_key: [u8; 32],
+    limit: u64,
+    count: AtomicU64,                 // The counter since last reset
+    last_reset: spin::Mutex<Instant>, // The time last reset was performed on this rate limiter
+}
+
+impl RateLimiter {
+    pub fn new(public_key: &X25519PublicKey, limit: u64) -> Self {
+        RateLimiter {
+            nonce_key: make_array(X25519SecretKey::new().as_bytes()), // A hack to get 32 random bytes
+            secret_key: make_array(X25519SecretKey::new().as_bytes()),
+            start_time: Instant::now(),
+            nonce_ctr0: AtomicU64::new(0),
+            nonce_ctr1: AtomicU64::new(0),
+            mac1_key: Blake2s::new_hash()
+                .hash(LABEL_MAC1)
+                .hash(public_key.as_bytes())
+                .finalize(),
+            mac2_key: Blake2s::new_hash()
+                .hash(LABEL_COOKIE)
+                .hash(public_key.as_bytes())
+                .finalize(),
+            limit,
+            count: AtomicU64::new(0),
+            last_reset: spin::Mutex::new(Instant::now()),
+        }
+    }
+
+    /// Reset packet count (ideally should be called with a period of 1 second)
+    pub fn reset_count(&self) {
+        // The rate limiter is not very accurate, but at the scale we care about it doesn't matter much
+        let current_time = Instant::now();
+        let mut last_reset_time = self.last_reset.lock();
+        if current_time.duration_since(*last_reset_time).as_secs() >= RESET_PERIOD {
+            self.count.store(0, Ordering::SeqCst);
+            *last_reset_time = current_time;
+        }
+    }
+
+    // Compute the correct cookie value based on the current secret value and the source IP
+    // adjustement can be used to compute past values if desired (for example using adjustment = 1 will
+    // return the cookie using the previous secret, and using adjustment = 2 will return the one before
+    // that).
+    fn current_cookie(&self, addr: IpAddr, adjustment: u64) -> Cookie {
+        let mut cookie = [0u8; COOKIE_SIZE];
+        let mut secret = [0u8; 16];
+        let mut addr_bytes = [0u8; 16];
+
+        match addr {
+            IpAddr::V4(a) => addr_bytes[..4].copy_from_slice(&a.octets()[..]),
+            IpAddr::V6(a) => addr_bytes[..].copy_from_slice(&a.octets()[..]),
+        }
+
+        // The current cookie for a given IP is the MAC(responder.changing_secret_every_two_minutes, initiator.ip_address)
+        // First we derive the secret from the current time, the value of cur_counter would change with time.
+        let cur_counter = Instant::now().duration_since(self.start_time).as_secs() / COOKIE_REFRESH;
+        assert_eq!(
+            COOKIE_SIZE,
+            ChaCha20Poly1305::new_aead(&self.secret_key[..]).seal_wg(
+                cur_counter - adjustment,
+                &addr_bytes[..],
+                &[],
+                &mut secret[..],
```

Comment:
Why the additional encryption, why is hashing not sufficient? E.g.
```
Keyed-Blake2s(secret_key, time / COOKIE_REFRESH || source_address)
```

Reply:
Because the spec says the cookie is MAC(responder.changing_secret_every_two_minutes, initiator.ip_address) I feel it is better to use the aead to derive the changing secret, and then use mac as stated.

Reply:
I do not see why the proposed computation would be better:
```
secret = ChaCha20-Poly1305(secret_key, time / COOKIE_REFRESH, aad=source_address, plaintext="")
current_cookie = Keyed-Blake2s(secret, source_address)
```

The computation occurs with exactly the same data. Since you do not need decryption, use of ChaCha20-Poly1305 seems a little strange. If you want to separate the key from the data, you could also use `secret = Keyed-Blake2s(secret_key, time / COOKIE_REFRESH)`, but again I see no advantages.

BTW, use of an empty plaintext implies that the source_address only affects the 128-bit auth tag at the end. There is no ciphertext either. You might as well have called Poly1305 (or any other hash function) directly.

Reply:
I don't think this construction is stronger either, just that I didn't want to diverge from the spec too much.

Using ChaCha as a sort of KDF is acceptable. All I need are random bytes, and the 16 byte tag does the job. Anyway, I will remove the extra KDF step.